# 1  How to use the ‘groupby()’ function in pandas

Welcome to this hands on about grouping data using pandas. This notebook presents the practical side of the concepts explained in the video. The dataset used in this notebook can be downloaded from [here](https://www.kaggle.com/datasets/subhaskumarray/netflixoriginals). This hands on covers the following topics:

• Split- Apply- Combine functionality

• Filtering the groups

• Data Transformation

Learners are expected to run this notebook on their own systems. Some of the output has not beenshown to keep the notebook precise.

# 1.1  Import pandas and load the data

The DataFrame here is named as movies_info. The downloaded dataset is saved in the present working directory.

In [1]:
pip install pandas

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.


In [2]:
import pandas as pd

In [3]:
movies_info = pd.read_csv('NetflixOriginals.csv', encoding='ISO-8859-1')

# 1.2  Understand the dataset

Let us begin by understanding the data using the commonly used methods.

In [4]:
movies_info.head()

Title                  Genre           Premiere  Runtime  \
0  Enter the Anime            Documentary     August 5, 2019       58   
1      Dark Forces               Thriller    August 21, 2020       81   
2          The App  Science fiction/Drama  December 26, 2019       79   
3   The Open House        Horror thriller   January 19, 2018       94   
4      Kaali Khuhi                Mystery   October 30, 2020       90   

   IMDB Score          Language  
0         2.5  English/Japanese  
1         2.6           Spanish  
2         2.6           Italian  
3         3.2           English  
4         3.4             Hindi

In [5]:
len(movies_info)

584

The dataset has 6 columns and 584 rows. The columns have features about movies such as theTitle, Genre, etc. Let us begin by forming groups from this data for further analysis.

# 1.3  Split

The first step is to split the dataset where grouping is done based on the values in a particular column.

# 1.3.1  Grouping data based on the value of a column

As we know that groupby() can form groups based on the value of a column in DataFrame. We can see this by forming groups based on Genre column. This grouping would provide information about the movies in each genre.

Find out how many movies of each genre exists in the DataFrame

In [6]:
grouped = movies_info.groupby('Genre')

Groups have been formed and stored in ‘grouped’ variable as a result of running the previous cell.We can see what groups have been formed by using .groups as shown below. Also, notice the alphabetical order of the groups obtained.

A snippet of the output has been shown below. (This is not the complete output).

{‘Action’: [5, 39, 50, 153, 220, 277, 372], ‘Action comedy’: [13, 61, 155, 205, 257], ‘Actionthriller’: [318], ‘Action-adventure’: [507], ‘Action-thriller’: [55, 315, 450], ‘Action/Comedy’: [99],‘Action/Science fiction’: [104], ‘Adventure’: [232, 337], ‘Adventure-romance’: [227], ‘Adven-ture/Comedy’: [122], ‘Aftershow / Interview’: [66, 407, 434, 439, 451, 528], ‘Animated musicalcomedy’: [366], ‘Animation’: [64, 253, 452, 453, 506], ‘Animation / Comedy’: [446],

In [7]:
grouped.groups

{'Action': [5, 39, 50, 153, 220, 277, 372], 'Action comedy': [13, 61, 155, 205, 257], 'Action thriller': [318], 'Action-adventure': [507], 'Action-thriller': [55, 315, 450], 'Action/Comedy': [99], 'Action/Science fiction': [104], 'Adventure': [232, 337], 'Adventure-romance': [227], 'Adventure/Comedy': [122], 'Aftershow / Interview': [66, 407, 434, 439, 451, 528], 'Animated musical comedy': [366], 'Animation': [64, 253, 452, 453, 506], 'Animation / Comedy': [446], 'Animation / Musicial': [309], 'Animation / Science Fiction': [537], 'Animation / Short': [325, 417, 513, 560], 'Animation/Christmas/Comedy/Adventure': [573], 'Animation/Comedy/Adventure': [314], 'Animation/Musical/Adventure': [306], 'Animation/Superhero': [51], 'Anime / Short': [40], 'Anime/Fantasy': [364], 'Anime/Science fiction': [323, 368], 'Anthology/Dark comedy': [543], 'Biographical/Comedy': [391], 'Biopic': [166, 207, 234, 239, 245, 255, 420, 449, 499], 'Black comedy': [123, 126], 'Christian musical': [141], 'Christmas

The output is a dictionary which shows the results of grouping as a key value pair, where,

•key is the ‘Genre’ or the group name;

•value is a list of the row labels of the group members.

as shown in the below output snippet. The complete output has not been shown to keep this document precise.

You can also see how many groups have been obtained by using len() on the dictionary obtained above.

In [8]:
len(grouped.groups)

115

There are 115 groups formed. The number of groups formed should be equal to the number of unique values in the Genre column since each unique genre would form a group of its own.

In [9]:
movies_info['Genre'].nunique()

115

You can also see the size of each group by using .size(). This shows the number of movies belonging to each genre.

In [10]:
grouped.size()

Genre
Action              7
Action comedy       5
Action thriller     1
Action-adventure    1
Action-thriller     3
                   ..
War                 2
War drama           2
War-Comedy          1
Western             3
Zombie/Heist        1
Length: 115, dtype: int64

# 1.4  Grouping data based on the value of multiple columns

Groups can also be created by using multiple columns. ### Find out how many movies belonging to each language are there in each genre.

In [11]:
grouped_1 = movies_info.groupby(['Genre', 'Language'])

Task: Also try what happens when the order of columns is reversed when passing multiple column names to groupby().

Let us have a look at the results of grouping using .groups This will output all the groups formed as we saw in the previous case. Part of the output has been shown below:

{('Action', 'English'): [153, 220, 277, 372], ('Action', 'French'): [39, 50], ('Action', 'Hindi'): [5], ('Action comedy', 'English'): [61, 155, 205, 257], ('Action comedy', 'Malay'): [13], ('Action thriller', 'English'): [318], ('Action-adventure', 'English/Korean'): [507], ('Action-thriller', 'English'): [55, 315], ('Action-thriller', 'Indonesian'): [450],

In [15]:
grouped_1.groups

{('Action', 'English'): [153, 220, 277, 372], ('Action', 'French'): [39, 50], ('Action', 'Hindi'): [5], ('Action comedy', 'English'): [61, 155, 205, 257], ('Action comedy', 'Malay'): [13], ('Action thriller', 'English'): [318], ('Action-adventure', 'English/Korean'): [507], ('Action-thriller', 'English'): [55, 315], ('Action-thriller', 'Indonesian'): [450], ('Action/Comedy', 'English'): [99], ('Action/Science fiction', 'English'): [104], ('Adventure', 'English'): [232, 337], ('Adventure-romance', 'English'): [227], ('Adventure/Comedy', 'English'): [122], ('Aftershow / Interview', 'English'): [66, 407, 434, 439, 451, 528], ('Animated musical comedy', 'English'): [366], ('Animation', 'English'): [64, 253, 452, 453, 506], ('Animation / Comedy', 'English'): [446], ('Animation / Musicial', 'English'): [309], ('Animation / Science Fiction', 'English'): [537], ('Animation / Short', 'English'): [325, 417, 513, 560], ('Animation/Christmas/Comedy/Adventure', 'English'): [573], ('Animation/Comedy

Let us find out the size of each group formed by using .size()

In [16]:
grouped_1.size()

Genre          Language    
Action         English         4
               French          2
               Hindi           1
Action comedy  English         4
               Malay           1
                              ..
War drama      English/Akan    1
War-Comedy     English         1
Western        English         2
               Portuguese      1
Zombie/Heist   English         1
Length: 204, dtype: int64

# 1.5  Selecting a group and forming subsets of data

# 1.5.1  Form a subset of the dataset containing only comedy movies

From the groups formed above,any group may be selected by using .get_group() and passing the group name. For example, the groups formedby using genre ‘Comedy’ canbe obtained by running the below cell.

In [19]:
grouped.get_group('Comedy')

Title   Genre  \
6                                   Leyla Everlasting  Comedy   
9                                  Sardar Ka Grandson  Comedy   
18                          What Happened to Mr. Cha?  Comedy   
22                                         Sextuplets  Comedy   
27                                   Seriously Single  Comedy   
28              The Misadventures of Hedi and Cokeman  Comedy   
29                                   5 Star Christmas  Comedy   
33   Porta dos Fundos: The First Temptation of Christ  Comedy   
35                                        The Players  Comedy   
47                                        Take the 10  Comedy   
48                                     The Main Event  Comedy   
54                                      Airplane Mode  Comedy   
65                                       #REALITYHIGH  Comedy   
69                                 Father of the Year  Comedy   
72                                   Girlfriend's Day  Comedy   
73                  Handsome: A Netflix Mystery Movie  Comedy   
74                                    Hubie Halloween  Comedy   
75                                              Ibiza  Comedy   
77                                       Sandy Wexler  Comedy   
83                                        The Week Of  Comedy   
85                                     Back to School  Comedy   
96                                 Despite Everything  Comedy   
103                                             Naked  Comedy   
110                                     Get the Grift  Comedy   
112                                      House Arrest  Comedy   
119                                      Step Sisters  Comedy   
124                                     Unicorn Store  Comedy   
125                                      Wine Country  Comedy   
127                                     Brahman Naman  Comedy   
138                                     The Sleepover  Comedy   
149                                         Lionheart  Comedy   
158                                   The Wrong Missy  Comedy   
160                                           Yes Day  Comedy   
167                                         Candy Jar  Comedy   
184                                   The After Party  Comedy   
221                      Between Two Ferns: The Movie  Comedy   
229                                       Like Father  Comedy   
231                                         Otherhood  Comedy   
249                     Have You Ever Seen Fireflies?  Comedy   
261                                        Win It All  Comedy   
271                                      Get the Goat  Comedy   
278               Porta dos Fundos: The Last Hangover  Comedy   
327                                        Chopsticks  Comedy   
342                      The Incredible Jessica James  Comedy   
377                            Just Another Christmas  Comedy   
397                              Long Live Brij Mohan  Comedy   
399                                     Death to 2020  Comedy   
447                                       Rose Island  Comedy   
492                           The 40-Year-Old Version  Comedy   

               Premiere  Runtime  IMDB Score    Language  
6      December 4, 2020      112         3.7     Turkish  
9          May 18, 2021      139         4.1       Hindi  
18      January 1, 2021      102         4.3      Korean  
22      August 16, 2019       99         4.4     English  
27        July 31, 2020      107         4.5     English  
28    February 10, 2021       99         4.5      French  
29     December 7, 2018       95         4.6     Italian  
33     December 3, 2019       46         4.6  Portuguese  
35        July 15, 2020       88         4.6     Italian  
47     January 20, 2017       80         4.8     English  
48       April 10, 2020      101         4.8     English  
54     January 23, 2020       96         5.0  Portuguese  
65    September 8, 2017       

The output will be a 49 row DataFrame consisting of all the columns which have the genre ‘Comedy’

# 1.6  Iterating through groups

To see all the groups formed as a result of groupby(), for loop can be used on the grouped data.The first group is shown below as an example. You would see 115 such groups. A sample output after one iteration has been shown below:

In [22]:
for name, group in grouped:
    print("Group Name:", name)
    print(group[['Title','Genre']],'\n')
    print('_____________________________End of group____________________________')

Group Name: Action
               Title   Genre
5              Drive  Action
39        Sentinelle  Action
50   Earth and Blood  Action
153      Point Blank  Action
220    6 Underground  Action
277            Polar  Action
372       Extraction  Action 

_____________________________End of group____________________________
Group Name: Action comedy
                                         Title          Genre
13                          All Because of You  Action comedy
61                             Coffee & Kareem  Action comedy
155                                The Do-Over  Action comedy
205  True Memoirs of an International Assassin  Action comedy
257                       Spenser Confidential  Action comedy 

_____________________________End of group____________________________
Group Name: Action thriller
        Title            Genre
318  Wheelman  Action thriller 

_____________________________End of group____________________________
Group Name: Action-adventure
    Title       

Let us also use for loop for the group obtained using 2 columns and see the result. The first group obtained after running the for loop is shown below.

In [25]:
for name, group in grouped_1:
    print("Group Name:", name)
    print(group, '\n')
    print('______________________________________________End of group__________________________________________')

Group Name: ('Action', 'English')
             Title   Genre           Premiere  Runtime  IMDB Score Language
153    Point Blank  Action      July 12, 2019       86         5.7  English
220  6 Underground  Action  December 13, 2019      128         6.1  English
277          Polar  Action   January 25, 2019      118         6.3  English
372     Extraction  Action     April 24, 2020      117         6.7  English 

______________________________________________End of group__________________________________________
Group Name: ('Action', 'French')
              Title   Genre        Premiere  Runtime  IMDB Score Language
39       Sentinelle  Action   March 5, 2021       80         4.7   French
50  Earth and Blood  Action  April 17, 2020       80         4.9   French 

______________________________________________End of group__________________________________________
Group Name: ('Action', 'Hindi')
   Title   Genre          Premiere  Runtime  IMDB Score Language
5  Drive  Action  November 1

You can also see the first entry in each group by using .first().

In [26]:
grouped.first()

Title           Premiere  Runtime  IMDB Score  \
Genre                                                                          
Action                         Drive   November 1, 2019      147         3.5   
Action comedy     All Because of You    October 1, 2020      101         4.2   
Action thriller             Wheelman   October 20, 2017       82         6.4   
Action-adventure                Okja      June 28, 2017      121         7.3   
Action-thriller          How It Ends      July 13, 2018      113         5.0   
...                              ...                ...      ...         ...   
War                      Sand Castle     April 21, 2017      113         6.3   
War drama                Da 5 Bloods      June 12, 2020      155         6.5   
War-Comedy               War Machine       May 26, 2017      122         6.0   
Western             The Ridiculous 6  December 11, 2015      119         4.8   
Zombie/Heist        Army of the Dead       May 21, 2021      148         5.9   

                        Language  
Genre                             
Action                     Hindi  
Action comedy              Malay  
Action thriller          English  
Action-adventure  English/Korean  
Action-thriller          English  
...                          ...  
War                      English  
War drama                English  
War-Comedy               English  
Western                  English  
Zombie/Heist             English  

[115 rows x 5 columns]

Similarly, last entry in each group can be obtained using .last()

In [27]:
grouped.last()

Title           Premiere  Runtime  \
Genre                                                                        
Action                              Extraction     April 24, 2020      117   
Action comedy             Spenser Confidential      March 6, 2020      111   
Action thriller                       Wheelman   October 20, 2017       82   
Action-adventure                          Okja      June 28, 2017      121   
Action-thriller         The Night Comes for Us   October 19, 2018      121   
...                                        ...                ...      ...   
War                    The Siege of Jadotville    October 7, 2016      108   
War drama                  Beasts of No Nation   October 16, 2015      136   
War-Comedy                         War Machine       May 26, 2017      122   
Western           The Ballad of Buster Scruggs  November 16, 2018      132   
Zombie/Heist                  Army of the Dead       May 21, 2021      148   

                  IMDB Score        Language  
Genre                                         
Action                   6.7         English  
Action comedy            6.2         English  
Action thriller          6.4         English  
Action-adventure         7.3  English/Korean  
Action-thriller          7.0      Indonesian  
...                      ...             ...  
War                      7.2         English  
War drama                7.7    English/Akan  
War-Comedy               6.0         English  
Western                  7.3         English  
Zombie/Heist             5.9         English  

[115 rows x 5 columns]

If you want to see the Nth entry in each group, you can use the .nth() function. The code shown below outputs the entry at row index 2.

In [29]:
grouped.nth(2)

Title  \
Genre                                                                         
Action                                                      Earth and Blood   
Action comedy                                                   The Do-Over   
Action-thriller                                      The Night Comes for Us   
Aftershow / Interview     I'm No Longer Here: A Discussion with Guillerm...   
Animation                                                Angela's Christmas   
Animation / Short                                   Sitara: Let Girls Dream   
Biopic                                                               Sergio   
Comedy                                            What Happened to Mr. Cha?   
Comedy-drama                                                 The Last Laugh   
Concert Film                    Barbra: The Music, The Mem'ries, The Magic!   
Crime drama                                                      Rogue City   
Documentary                                                  After the Raid   
Drama                                                           The Binding   
Horror                                   Nobody Sleeps in the Woods Tonight   
Horror thriller                                               Gerald's Game   
One-man show                                        Springsteen on Broadway   
Psychological thriller                        I'm Thinking of Ending Things   
Romance                                                              Tramps   
Romantic comedy                                                Squared Love   
Romantic drama                                        Christmas Inheritance   
Science fiction                                            The Midnight Sky   
Science fiction/Drama                                         The Discovery   
Science fiction/Thriller                                               iBoy   
Sports-drama                                                    First Match   
Thriller                                                   Secret Obsession   
Variety show                                Sarah Cooper: Everything's Fine   
Western                                        The Ballad of Buster Scruggs   

                                    Premiere  Runtime  IMDB Score    Language  
Genre                                                                          
Action                        April 17, 2020       80         4.9      French  
Action comedy                   May 27, 2016      108         5.7     English  
Action-thriller             October 19, 2018      121         7.0  Indonesian  
Aftershow / Interview       November 3, 2020       14         7.0     English  
Animation                  November 30, 2018       30         7.1     English  
Animation / Short              March 8, 2020       15         7.3     English  
Biopic                        April 17, 2020      118         6.1     English  
Comedy                       January 1, 2021      102         4.3      Korean  
Comedy-drama                January 11, 2019       98         5.6     English  
Concert Film               November 22, 2017      108         7.5     English  
Crime drama                 October 30, 2020      116         6.1      French  
Documentary                December 19, 2019       25         4.3     Spanish  
Drama                        October 2, 2020       93         4.7     Italian  
Horror                      October 28, 2020      103         4.8      Polish  
Horror thriller           September 29, 2017      103         6.5     English  
One-man show               December 16, 2018      153         8.5     English  
Psychological thriller     September 4, 2020      134         6.6     English  
Romance                       April 21, 2017       83         6.5     English  
Romantic comedy            February 11, 2021      102         5.0      Polish  
Romantic drama             December 15, 2017      104         5.7     English  
Science fiction   

Let us now see the first entry in grouped_1.

In [30]:
grouped_1.first()

Title           Premiere  Runtime  \
Genre         Language                                                        
Action        English               Point Blank      July 12, 2019       86   
              French                 Sentinelle      March 5, 2021       80   
              Hindi                       Drive   November 1, 2019      147   
Action comedy English           Coffee & Kareem      April 3, 2020       88   
              Malay          All Because of You    October 1, 2020      101   
...                                         ...                ...      ...   
War drama     English/Akan  Beasts of No Nation   October 16, 2015      136   
War-Comedy    English               War Machine       May 26, 2017      122   
Western       English          The Ridiculous 6  December 11, 2015      119   
              Portuguese             The Killer  November 10, 2017       99   
Zombie/Heist  English          Army of the Dead       May 21, 2021      148   

                            IMDB Score  
Genre         Language                  
Action        English              5.7  
              French               4.7  
              Hindi                3.5  
Action comedy English              5.1  
              Malay                4.2  
...                                ...  
War drama     English/Akan         7.7  
War-Comedy    English              6.0  
Western       English              4.8  
              Portuguese           6.1  
Zombie/Heist  English              5.9  

[204 rows x 4 columns]

# 1.7  Sorting when using groupby

As we have seen earlier, groups are listed in alphabetical order from A to Z. If sort is set to ‘False’,then no ordering will take place when forming groups. This can be seen in the below cells.

In [32]:
grouped_2 = movies_info.groupby('Genre', sort = False)

You can see the result of grouping using .groups. Notice how the group names are arranged in order of appearance in the dataset.

In [33]:
grouped_2.groups

{'Documentary': [0, 10, 15, 20, 30, 36, 80, 111, 140, 151, 152, 199, 202, 223, 228, 259, 260, 263, 275, 286, 290, 292, 295, 301, 303, 304, 308, 310, 312, 316, 320, 324, 326, 329, 334, 339, 351, 353, 355, 356, 357, 359, 363, 367, 370, 371, 373, 375, 378, 384, 387, 392, 393, 394, 396, 400, 401, 402, 403, 404, 410, 412, 414, 415, 416, 419, 424, 425, 426, 427, 432, 433, 435, 436, 437, 440, 441, 442, 443, 444, 445, 455, 456, 458, 459, 460, 462, 463, 464, 465, 466, 468, 469, 471, 472, 473, 476, 479, 480, 482, ...], 'Thriller': [1, 14, 21, 23, 25, 43, 45, 57, 60, 86, 88, 89, 100, 109, 117, 130, 137, 154, 159, 172, 192, 201, 242, 251, 252, 254, 288, 298, 313, 332, 398, 413, 509], 'Science fiction/Drama': [2, 38, 283], 'Horror thriller': [3, 19, 331], 'Mystery': [4, 196], 'Action': [5, 39, 50, 153, 220, 277, 372], 'Comedy': [6, 9, 18, 22, 27, 28, 29, 33, 35, 47, 48, 54, 65, 69, 72, 73, 74, 75, 77, 83, 85, 96, 103, 110, 112, 119, 124, 125, 127, 138, 149, 158, 160, 167, 184, 221, 229, 231, 249, 2

# 1.8  Apply

After splitting the data, the next step is to apply functions to the groups so formed. These function scan be applied in various ways as can be seen in the below cells.

# 1.8.1  User defined functions

Find out the average Runtime for each genre of movies.

In [35]:
movies_info.groupby('Genre')[['Runtime']].mean()

Runtime
Genre                       
Action            108.000000
Action comedy     101.200000
Action thriller    82.000000
Action-adventure  121.000000
Action-thriller   119.666667
...                      ...
War               110.500000
War drama         145.500000
War-Comedy        122.000000
Western           116.666667
Zombie/Heist      148.000000

[115 rows x 1 columns]

Find out the size of each group after grouping.

In [36]:
movies_info.groupby(['Genre','Language']).size()

Genre          Language    
Action         English         4
               French          2
               Hindi           1
Action comedy  English         4
               Malay           1
                              ..
War drama      English/Akan    1
War-Comedy     English         1
Western        English         2
               Portuguese      1
Zombie/Heist   English         1
Length: 204, dtype: int64

Find the minimum Score for each genre

In [37]:
movies_info.groupby('Genre')[['IMDB Score']].min()

IMDB Score
Genre                       
Action                   3.5
Action comedy            4.2
Action thriller          6.4
Action-adventure         7.3
Action-thriller          5.0
...                      ...
War                      6.3
War drama                6.5
War-Comedy               6.0
Western                  4.8
Zombie/Heist             5.9

[115 rows x 1 columns]

Find the number of movies in each language in the dataset

In [38]:
movies_info.groupby('Language').size()

Language
Bengali                         1
Dutch                           3
English                       401
English/Akan                    1
English/Arabic                  1
English/Hindi                   2
English/Japanese                2
English/Korean                  1
English/Mandarin                2
English/Russian                 1
English/Spanish                 5
English/Swedish                 1
English/Taiwanese/Mandarin      1
English/Ukranian/Russian        1
Filipino                        2
French                         20
Georgian                        1
German                          5
Hindi                          33
Indonesian                      9
Italian                        14
Japanese                        6
Khmer/English/French            1
Korean                          6
Malay                           1
Marathi                         3
Norwegian                       1
Polish                          3
Portuguese                     12
Spani

# 1.8.2  Working with apply()

apply() can be used to apply an operation to the groups. Some examples have been shown below.

Find the maximum Runtime for each genre using lambda function .Lambda function finds the maximum in each group formed by groupby().

In [41]:
movies_info.groupby('Genre')[['Runtime']].apply(lambda gr:gr.max())

Runtime
Genre                    
Action                147
Action comedy         111
Action thriller        82
Action-adventure      121
Action-thriller       125
...                   ...
War                   113
War drama             155
War-Comedy            122
Western               132
Zombie/Heist          148

[115 rows x 1 columns]

Find the maximum Runtime for each genre using the inbuilt function.

In [42]:
movies_info.groupby('Genre')[['Runtime']].apply(max)

Runtime
Genre                    
Action                147
Action comedy         111
Action thriller        82
Action-adventure      121
Action-thriller       125
...                   ...
War                   113
War drama             155
War-Comedy            122
Western               132
Zombie/Heist          148

[115 rows x 1 columns]

Find the maximum Runtime and IMDB Score for each genre using the user defined function.

In [44]:
def find_max(df):
    return df.max()

movies_info.groupby('Genre')[['Runtime', 'IMDB Score']].apply(find_max)

Runtime  IMDB Score
Genre                                
Action              147.0         6.7
Action comedy       111.0         6.2
Action thriller      82.0         6.4
Action-adventure    121.0         7.3
Action-thriller     125.0         7.0
...                   ...         ...
War                 113.0         7.2
War drama           155.0         7.7
War-Comedy          122.0         6.0
Western             132.0         7.3
Zombie/Heist        148.0         5.9

[115 rows x 2 columns]

# 1.8.3  Aggregation

agg() can be used to find summary statistics about the data such as count, minimum, maximumand mean in a single step.

Find out the count, minimum, maximum and average of the IMDB Score for eachgenre.

In [45]:
movies_info.groupby('Genre')['IMDB Score'].agg(['count','min','max','mean'])

count  min  max      mean
Genre                                      
Action                7  3.5  6.7  5.414286
Action comedy         5  4.2  6.2  5.420000
Action thriller       1  6.4  6.4  6.400000
Action-adventure      1  7.3  7.3  7.300000
Action-thriller       3  5.0  7.0  6.133333
...                 ...  ...  ...       ...
War                   2  6.3  7.2  6.750000
War drama             2  6.5  7.7  7.100000
War-Comedy            1  6.0  6.0  6.000000
Western               3  4.8  7.3  6.066667
Zombie/Heist          1  5.9  5.9  5.900000

[115 rows x 4 columns]

Find out the minimum and maximum duration and minimum and maximum ratings for each genre

In [46]:
movies_info.groupby('Genre').agg({'Runtime': ['max','min'],'IMDB Score': ['mean','max']})

Runtime      IMDB Score     
                     max  min       mean  max
Genre                                        
Action               147   80   5.414286  6.7
Action comedy        111   88   5.420000  6.2
Action thriller       82   82   6.400000  6.4
Action-adventure     121  121   7.300000  7.3
Action-thriller      125  113   6.133333  7.0
...                  ...  ...        ...  ...
War                  113  108   6.750000  7.2
War drama            155  136   7.100000  7.7
War-Comedy           122  122   6.000000  6.0
Western              132   99   6.066667  7.3
Zombie/Heist         148  148   5.900000  5.9

[115 rows x 4 columns]

Find out the average runtime and the maximum IMDB Score for each language

In [48]:
movies_info.groupby('Language').agg({'Runtime':'mean','IMDB Score':'max'})

Runtime  IMDB Score
Language                                          
Bengali                      41.000000         7.1
Dutch                        99.666667         7.1
English                      91.817955         9.0
English/Akan                136.000000         7.7
English/Arabic              114.000000         7.3
English/Hindi                32.500000         7.4
English/Japanese             89.000000         6.3
English/Korean              121.000000         7.3
English/Mandarin             59.000000         7.3
English/Russian              90.000000         7.3
English/Spanish              39.200000         7.3
English/Swedish              40.000000         6.5
English/Taiwanese/Mandarin   91.000000         6.5
English/Ukranian/Russian     91.000000         8.4
Filipino                     99.500000         5.5
French                       92.700000         6.9
Georgian                     23.000000         6.8
German                       99.600000         6.4
Hindi                       115.787879         7.6
Indonesian                  103.777778         7.0
Italian                      98.357143         7.3
Japanese                     99.333333         6.7
Khmer/English/French        136.000000         7.2
Korean                      115.833333         7.5
Malay                       101.000000         4.2
Marathi                     121.666667         7.2
Norwegian                    86.000000         5.1
Polish                       98.666667         5.7
Portuguese                   91.250000         8.6
Spanish                      93.387097         8.2
Spanish/Basque               89.000000         5.6
Spanish/Catalan             116.000000         6.4
Spanish/English              96.000000         7.3
Swedish                      86.000000         5.5
Tamil                       101.000000         7.2
Thai                        101.000000         5.7
Thia/English                 80.000000         6.7
Turkish                     101.800000         6.7

From this you can see that Hindi language movies have the longest Runtime and also the highest IMDB Score.

# 1.9  Transformation

While apply stage includes applying some function to the grouped data to find an overall summary for the group, transformation includes applying a function to each element in the group to modify it. For example, the Runtime is given in minutes and we can transform it to hours by dividing each value by 60. .transform can be used to perform such operations on the grouped data as shown below.

A new column is added to the DataFrame to hold the transformed values.

In [52]:
movies_info['Runtime_hrs']= movies_info.groupby('Genre')['Runtime'].transform(lambda x:x/60)

In [53]:
movies_info.head()

Title                  Genre           Premiere  Runtime  \
0  Enter the Anime            Documentary     August 5, 2019       58   
1      Dark Forces               Thriller    August 21, 2020       81   
2          The App  Science fiction/Drama  December 26, 2019       79   
3   The Open House        Horror thriller   January 19, 2018       94   
4      Kaali Khuhi                Mystery   October 30, 2020       90   

   IMDB Score          Language  Runtime_hrs  
0         2.5  English/Japanese     0.966667  
1         2.6           Spanish     1.350000  
2         2.6           Italian     1.316667  
3         3.2           English     1.566667  
4         3.4             Hindi     1.500000

Let us now replace the IMDB Score in each genre by the genre’s mean IMDB Score.

In [54]:
movies_info['IMDB Score adj'] = movies_info.groupby('Genre')['IMDB Score'].transform(lambda x:x.mean())

In [55]:
movies_info.head()

Title                  Genre           Premiere  Runtime  \
0  Enter the Anime            Documentary     August 5, 2019       58   
1      Dark Forces               Thriller    August 21, 2020       81   
2          The App  Science fiction/Drama  December 26, 2019       79   
3   The Open House        Horror thriller   January 19, 2018       94   
4      Kaali Khuhi                Mystery   October 30, 2020       90   

   IMDB Score          Language  Runtime_hrs  IMDB Score adj  
0         2.5  English/Japanese     0.966667        6.936478  
1         2.6           Spanish     1.350000        5.563636  
2         2.6           Italian     1.316667        4.533333  
3         3.2           English     1.566667        4.700000  
4         3.4             Hindi     1.500000        4.650000

# 1.10  Filter

After grouping the data, groups can be filtered by using some condition. For example, in the belowcell, Genre with a average IMDB Score higher than 4 have been filtered using .filter(

In [56]:
movies_info.groupby('Genre').filter(lambda x:x['IMDB Score'].mean()>=4)

Title                  Genre  \
0                                Enter the Anime            Documentary   
1                                    Dark Forces               Thriller   
2                                        The App  Science fiction/Drama   
3                                 The Open House        Horror thriller   
4                                    Kaali Khuhi                Mystery   
..                                           ...                    ...   
579        Taylor Swift: Reputation Stadium Tour           Concert Film   
580  Winter on Fire: Ukraine's Fight for Freedom            Documentary   
581                      Springsteen on Broadway           One-man show   
582    Emicida: AmarElo - It's All For Yesterday            Documentary   
583     David Attenborough: A Life on Our Planet            Documentary   

              Premiere  Runtime  IMDB Score                  Language  \
0       August 5, 2019       58         2.5          English/Japanese   
1      August 21, 2020       81         2.6                   Spanish   
2    December 26, 2019       79         2.6                   Italian   
3     January 19, 2018       94         3.2                   English   
4     October 30, 2020       90         3.4                     Hindi   
..                 ...      ...         ...                       ...   
579  December 31, 2018      125         8.4                   English   
580    October 9, 2015       91         8.4  English/Ukranian/Russian   
581  December 16, 2018      153         8.5                   English   
582   December 8, 2020       89         8.6                Portuguese   
583    October 4, 2020       83         9.0                   English   

     Runtime_hrs  IMDB Score adj  
0       0.966667        6.936478  
1       1.350000        5.563636  
2       1.316667        4.533333  
3       1.566667        4.700000  
4       1.500000        4.650000  
..           ...             ...  
579     2.083333        7.633333  
580     1.516667        6.936478  
581     2.550000        7.133333  
582     1.483333        6.936478  
583     1.383333        6.936478  

[582 rows x 8 columns]

# 1.11  Combine

In the final stage, let us see how the results of applying different operations to the grouped data are combined into a DataFrame. This is done automatically in most cases. For example, using agg() returns the summary of the data in the form of a DataFrame.

In [60]:
movies_combine = movies_info.groupby('Genre').agg({'Runtime': ['max','min'],'IMDB Score':['mean','max']})

In [58]:
type(movies_combine)

pandas.core.frame.DataFrame

You can further use this DataFrame for various purposes.

In [63]:
movies_combine.head()

Runtime      IMDB Score     
                     max  min       mean  max
Genre                                        
Action               147   80   5.414286  6.7
Action comedy        111   88   5.420000  6.2
Action thriller       82   82   6.400000  6.4
Action-adventure     121  121   7.300000  7.3
Action-thriller      125  113   6.133333  7.0

You can also create a DataFrame as shown below. The below cell summarises all the three stages:

•Split:movies_info.groupby(‘Genre’)

•Apply:movies_info.groupby(‘Genre’).size()

•Combine:pd.DataFrame(movies_info.groupby(‘Genre’).size(), columns= [‘Number’]

In [65]:
pd.DataFrame(movies_info.groupby('Genre').size(),columns = ['Number'])

Number
Genre                   
Action                 7
Action comedy          5
Action thriller        1
Action-adventure       1
Action-thriller        3
...                  ...
War                    2
War drama              2
War-Comedy             1
Western                3
Zombie/Heist           1

[115 rows x 1 columns]